VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitWebviewFragment.java

```diff
@@ -71,12 +77,38 @@ public void onActivityCreated(Bundle savedInstanceState) {
             @Override
             public void onReceivedError(WebView view, int errorCode,
                                         String description, String failingUrl) {
-                int error = errorCode;//for testing only
+                if ( callback != null )
+                    callback.finishProcess();
+                if ( errorCode == HttpStatus.SC_FORBIDDEN || errorCode == HttpStatus.SC_UNAUTHORIZED ){
+                    tryToLoadWebview(view);
+                }
+                super.onReceivedError(view, errorCode, description, failingUrl);
             }
             public void onPageFinished(WebView view, String url) {
+                if ( callback != null )
+                    callback.finishProcess();
+
+                String sessionIdRefreshUrl = Api.getSessionTokenExchangeUrl();
+                if ( url.equals(sessionIdRefreshUrl) ){
+                    String cookies = CookieManager.getInstance().getCookie(url);
+                    String[] cookieValues=cookies.split(";");
+                    for (int i = 0; i < cookieValues.length; i++) {
+                        String[] split = cookieValues[i].split("=");
+                        if ( split[0].trim().equals(PrefManager.Key.SESSION_ID) ){
+                            PrefManager pref = new PrefManager(getActivity(), PrefManager.Pref.LOGIN);
+                            pref.put(PrefManager.Key.AUTH_ASSESSMENT_SESSION_ID, split[1]);
+                            //we can not get the expiration date from cookie, so just set it to expire for one day
+                            pref.put(PrefManager.Key.AUTH_ASSESSMENT_SESSION_EXPIRATION, new Date().getTime() + 3600 * 24 * 1000 );
```

Comment:
@aleffert I print out the cookie content during debugging, there is not expiration date field.